Use Capstone for x86 instruction decoding

hainest opened the pull request:
This replaces #638.

This is still a work in progress, but I want to get the code out for review. There are several items I want to check on before even considering merging:

- [x] Verify all decode cases from `InstructionDecoder_x86` have been captured
- [x] Check completeness and accuracy of `decode_reg`
- [x] Check completeness and accuracy of `decode_imm`
- [x] Check completeness and accuracy of `decode_mem`
- [x] Handle segment registers in `decode_mem`
- [x] Replace calculations in `Operation::SetUpNonOperandData` (at least for x86)
- [x] Fix error handling in `decode_mem`
- [x] Update container builds to include Capstone source
- [ ] Check that the description of `Instruction::getWriteSet/getReadSet` is correct

Big items as separate PRs
- [ ] Get rid of old decode tables
- [ ] Check consistency of register usages in dataflowAPI, parseAPI, etc. For example, popf{d,q} are not uniformly checked in switches.
- [ ] Try to replace manual register checks with generalized concepts and put them into new methods in the Instruction class

**decoder.C**
- [ ] x86_decoder::decode_operands
  - Don't need to check for endbr32/64 because no operands have been set on the `insn` object (it's also const here).

**Operation.C**
- [x] Rename file
  -  File is no longer needed
- [x] Why is Operation_impl inheriting from boost::lockable_adapter?
  - It shouldn't.
- [x] Remove `Operation(NS_X86::ia32_entry*, ...)`
  - Fixed by #2169
- [x] ~~Why can't `SetUpNonOperandData` be done in the ctor?~~ It's not needed. Removed.
  - this would get rid of mutables and let the accessors be 'const' again
  - Why is it x86 only?
  - read{write}Flags checks are incomplete.
- [x] `Operation(entryID id,...)` should check all architecture values in switch; no default
  - `archDecodedFrom` is no longer used
- [x] ~~Move once_flag ito pImpl~~ It's not needed. Removed.
  - Get rid of SMFs.
- [ ] Use updateMnemonic throughout Dyninst instead of the idiom 'Instruction::getOperation().mnemonic = ...'
  - Probably make them friends and move updateMnemonic to private
  - Then can remove non-const Instruction::getOperation

**instructionAPI/h/InstructionCategories.{h,C}**
- [ ] Update entryToCategory after getting new Capstone mnemonics
- [ ] entryToCategory
  - This is just buggered
- [ ] makeReturnExpression should use u32/u64 depending on `arch_decoded_from`.
  - This makes its use in x86_decoder::decode_operands consistent with the expr created in InstructionDecoder_x86::decodeOperands:1649.

**instruction.C**
- [ ] Remove DEBUG_INSN_ALLOCATIONS
- [ ] makeReturnExpression should be synchronized with InstructionDecoderImpl::makeRegisterExpression

**x86{_64}_regs.h**
- [ ] Remove tr* registers (they only exist for Dyninst's internal decoder)
- [ ] Possibly remove MISC registers

**entryIDs**
- [ ] popf{d,q} are not uniformly checked in switches
- [ ] collapse all e_ret* into just e_ret
  - Capstone will generate aliases for them, but we should not treat them distinctly
  - For example, convertOpcodes.C:159 should go away
- [ ] Remove e_fp_generic
- [ ] Remove e_int80
- [ ] cwde and cwdq should alias cbw, but Dyninst has separate semantics for them (see x86_64InstructionSemantics.h) even though it doesn't have separate decode table entries for them.

**BoundFactCalculator.C**
- [ ] 457: Assumes all SETxx entry ids are contiguous

**x86{_64}_regs.h**
- [ ] Check which instructions read/write IOPL bit of EFLAGS

**instructionAPI/InstructionTest.C**
- [ ] Keep or throw away?
  - throw it away. It has fixed mnemonics that won't necessarily match capstone's.

---
Lower priority items:
- [ ] Handle implicit x87 control/state registers

Fixes #1881 (modulo missing reg read)

hainest committed:
Add Capstone->Dyninst register translation

hainest committed:
Add Capstone->Dyninst opcode translation

kupsch reviewed (approved):
OK.  Just one incorrect comment.

bbiiggppiigg commented on instructionAPI/src/x86/decoder.C:
```diff
@@ -0,0 +1,359 @@
+#include <map>
+#include <vector>
+
+namespace di = Dyninst::InstructionAPI;
```
I'm not entirely sure.
Is this a good practice to shorten namespace like this?
Or is it better to keep them as Dyninst::instructionAPI::Result_Type?
Or maybe these are not needed cause I assume decoder is in the instructionAPI namespace?

hainest replied:
It's all good. I just wouldn't do this in the global namespace in a public header.

bbiiggppiigg reviewed (approved):
I left a comment on the use of shortened namespace.

hainest commented on instructionAPI/src/x86/register-xlat.C:
```diff
@@ -0,0 +1,531 @@
+          case X86_REG_FP6:
+          case X86_REG_FP7:
+
+          /* AMX registers */
```
Fixed.

hainest commented:
@kupsch @wxrdnx I merged master. You should be able to fast forward, unless you've made conflicting local changes. Let me know if you have any issues.

hainest committed:
Update syscalls to handle corrected AST

The new x86 instruction decoder correctly creates AST for the cases like
`call gs:[0x10]` to be the same as for `call [Base + 0x10]`. Moreover,
the new decoder handles generic segment-based addressing allowed in
64-bit mode like `call [Seg + Base + (Index * Scale) + Disp]` which is
not a system call.

hainest committed:
Add register_set

This is a collection of MachRegisters with set operations but without
any ordering constraints.

hainest committed:
x87 tests

hainest committed:
arithmetic tests

hainest committed:
branches tests

hainest committed:
data transfer tests

hainest committed:
sse tests

hainest committed:
stack tests

hainest committed:
Add 'isConditional' test in cft tests

hainest committed:
string tests

hainest committed:
sysctl tests

hainest committed:
Decode explicit register operands

hainest committed:
Decode implicit memory operands

Capstone doesn't track these at all. The additional memory op for push
instructions was taken directly from Operation::setupNonOperandData.

hainest committed:
Decode implicit flags for interrupt instructions

Capstone doesn't fully track eflags for interrupt instructions.

hainest committed:
Remove old x86 decoder

hainest committed:
Add stub replacement for x86 decoder

hainest committed:
Implement x86_decoder::decode

hainest committed:
Implement skeleton for x86_decoder::decode_operands

hainest committed:
Decode explicit memory operands

hainest committed:
Decode implicit operands

hainest committed:
call tests

hainest committed:
Remove Operation::implicit{Reads,Writes}

All implicit operands are now accounted for in the decoder.

hainest committed:
Remove Operation::is{Read,Written}

All implicit operands are now accounted for in the decoder.

hainest committed:
Remove Operation::getImplicitMem{Reads,Writes}

All implicit operands are now accounted for in the decoder.

hainest committed:
Remove unused Operation::addrWidth

hainest committed:
Cleanup, inline Operation ctor

hainest committed:
Remove x86 entry lookup in Operation::format

The decoder always populates the mnemonic, so this code will never be
used.

hainest committed:
Operation: inline all member functions

hainest committed:
Remove Operation.C

All of it has been removed or inlined into its header.

hainest committed:
Decode explicit immediate operands

hainest committed:
Decode instruction categories

hainest committed:
Use new categories in Instruction

hainest committed:
Remove Operation::archDecodedFrom

It's no longer needed.

hainest committed:
Only use explicit operand to make assignments for x86 pop

It is no longer guaranteed that the first operand coming from
getAllOperands is the destination. It _is_ guaranteed to be the first
explicit operand- because -it's the only explicit one.

hainest committed:
Do not build a jump table for an instruction without an assignment

This can happen when semantic information is incomplete for the
instruction.

hainest committed:
Merge in RISCV changes for latest Capstone

hainest committed:
Remove Operation::SetUpNonOperandData

All implicit operands are now accounted for in the decoder.

hainest committed:
Remove Operation SMFs

Now that the std::once_flag has been removed, the special member
functions (SMFs) are no longer needed. The compiler-generated ones
will work correctly.

hainest committed:
Remove DYNINST_ENABLE_CAPSTONE

It's always required now.

hainest committed:
Remove unneeded friend declarations in Operation
